jni/src/Foreign/JNI.cpphs

```diff
@@ -336,34 +340,40 @@ globalJVMLock :: RWLock
 globalJVMLock = unsafePerformIO newRWLock
 {-# NOINLINE globalJVMLock #-}
 
-jni_ok :: Int32
-jni_ok = [CU.pure| jint { JNI_OK } |]
-{-# NOINLINE jni_ok #-}
+throwIfNotOK_ :: HasCallStack => IO Int32 -> IO ()
+throwIfNotOK_ m = m >>= \case
+  rc
+    | rc == [CU.pure| jint { JNI_OK } |] -> return ()
+    | rc == [CU.pure| jint { JNI_EDETACHED } |] -> throwIO ThreadNotAttached
+    | otherwise -> throwIO $ JNIError (prettySrcLoc loc) rc
+  where
+    (_, loc):_ = getCallStack callStack
 
 attachCurrentThreadAsDaemon :: IO ()
-attachCurrentThreadAsDaemon = do
-    rc <- [CU.exp| jint {
+attachCurrentThreadAsDaemon = uninterruptibleMask $ \_ -> do
+    throwIfNotOK_
+      [CU.exp| jint {
         (*$(JavaVM* jvm))->AttachCurrentThreadAsDaemon($(JavaVM* jvm), &jniEnv, NULL)
       } |]
-    when (rc /= jni_ok) $
-      throwIO $ JNIError "attachCurrentThreadAsDaemon" rc
+    throwIfNotOK_
+      [CU.exp| jint {
+        (*$(JavaVM* jvm))->GetEnv($(JavaVM* jvm), (void**)&jniEnv, JNI_VERSION_1_6)
+      }|]
```

Comment:
`AttachCurrentThreadAsDaemon` already sets `jniEnv`, I don't understand what this call to `GetEnv` acomplishes.

Reply:
Ok good. I had not spotted that. So we can even do away with the `uninterruptibleMask`.

Reply:
Also, I don't get what `uninterruptibleMask` is for here.

Reply:
Right.

Reply:
well, to maintain the stated invariant in the face of `catch` you need to protect against async exceptions. Foreign calls are not normally interruptible, but here we had two. Another solution would have been to have just one.